// ace-master-with-pegjs/build/src/mode-pegjs.js
define("ace/mode/pegjs_highlight_rules",["require","exports","module","ace/lib/oop","ace/mode/text_highlight_rules"], function(require, exports, module) {
"use strict";

var oop = require("../lib/oop");
var TextHighlightRules = require("./text_highlight_rules").TextHighlightRules;

var PEGjsHighlightRules = function() {

    this.$rules = {
        start: [{
            include: "#commentBlock"
        }, {
            include: "#commentLine"
        }, {
            include: "#docCommentLine"
        }, {
            include: "#exprLabel"
        }, {
            include: "#pegOperator"
        }, {
            include: "#charSet"
        }, {
            include: "#literalDouble"
        }, {
            include: "#literalSingle"
        }, {
            token: [
                "entity.name.type",
                "source.pegjs.ruleDefinition",
                "comment.line.double-slash",
                "source.pegjs.ruleDefinition"
            ],
            regex: /^([a-zA-Z$_][a-zA-Z$_0-9]*)([\t ]*)(".*")([\t ]*=?)/
        }, {
            token: [
                "entity.name.type",
                "source.pegjs.ruleDefinition",
                "comment.line.double-slash",
                "source.pegjs.ruleDefinition"
            ],
            regex: /^([a-zA-Z$_][a-zA-Z$_0-9]*)([\t ]*)('.*')([\t ]*=?)/
        }, {
            token: [
                "entity.name.type",
                "source.pegjs.ruleDefinition"
            ],
            regex: /^([a-zA-Z_][a-zA-Z$_0-9]*)([\s\t]*=?)/
        }, {
            include: "#scriptblock"
        }],
        "#charSet": [{
            token: "keyword.operator",
            regex: /\[/,
            push: [{
                token: "keyword.operator",
                regex: /\]/,
                next: "pop"
            }, {
                defaultToken: "keyword.operator"
            }]
        }],
        "#commentBlock": [{
            token: "punctuation.definition.comment.pegjs",
            regex: /\/\*/,
            push: [{
                token: "punctuation.definition.comment.pegjs",
                regex: /\*\//,
                next: "pop"
            }, {
                defaultToken: "comment.block"
            }]
        }],
        "#commentLine": [{
            token: [
                "punctuation.definition.comment.js",
                "comment.line.double-slash.js"
            ],
            regex: /(\/\/)(.*$)/
        }],
        "#docCommentBlock": [{
            token: "punctuation.definition.comment.pegjs",
            regex: /\/\*\*(?!\/)/,
            push: [{
                token: "punctuation.definition.comment.pegjs",
                regex: /\*\//,
                next: "pop"
            }, {
                defaultToken: "comment.block.documentation.peg"
            }]
        }],
        "#exprLabel": [{
            token: "variable.parameter",
            regex: /[a-zA-Z$_][a-zA-Z$_0-9]*:/
        }],
        "#literalDouble": [{
            token: "punctuation.definition.string.begin.pegjs",
            regex: /"/,
            push: [{
                token: "punctuation.definition.string.end.pegjs",
                regex: /"/,
                next: "pop"
            }, {
                token: "constant.character.escape.pegjs",
                regex: /\\(?:x[\da-fA-F]{2}|[0-2][0-7]{,2}|3[0-6][0-7]|37[0-7]?|[4-7][0-7]?|.)/
            }, {
                defaultToken: "string.quoted.double.pegjs"
            }]
        }],
        "#literalSingle": [{
            token: "punctuation.definition.string.begin.pegjs",
            regex: /'/,
            push: [{
                token: "punctuation.definition.string.end.pegjs",
                regex: /'/,
                next: "pop"
            }, {
                token: "constant.character.escape.js",
                regex: /\\(?:x[\da-fA-F]{2}|[0-2][0-7]{,2}|3[0-6][0-7]?|37[0-7]?|[4-7][0-7]?|.)/
            }, {
                defaultToken: "string.quoted.single.js"
            }]
        }],
        "#pegOperator": [{
            token: "keyword.operator",
            regex: /[\/.+*?&!=]/
        }],
        "#rule": [{
            token: "entity.name.type",
            regex: /[a-zA-Z$_][a-zA-Z$_0-9]*/
        }],
        "#ruleDef": [{
            token: "entity.name.type",
            regex: /[_a-zA-Z$_][a-zA-Z$_0-9]*/
        }],
        "#scriptblock": [{
            token: "source.js.embedded.pegjs",
            regex: /\{/,
            push: [{
                token: "source.js.embedded.pegjs",
                regex: /\}/,
                next: "pop"
            }, {
                include: "#scriptblock"
            }, {
                include: "source.js"
            }, {
                defaultToken: "source.js.embedded.pegjs"
            }]
        }]
    }
    
    this.normalizeRules();
};

PEGjsHighlightRules.metaData = {
    fileTypes: ["pegjs"],
    foldingStartMarker: "/\\*\\*|\\{\\s*$",
    foldingStopMarker: "\\*\\*/|^\\s*\\}",
    name: "PEGjs",
    scopeName: "source.pegjs"
}


oop.inherits(PEGjsHighlightRules, TextHighlightRules);

exports.PEGjsHighlightRules = PEGjsHighlightRules;
});

define("ace/mode/folding/cstyle",["require","exports","module","ace/lib/oop","ace/range","ace/mode/folding/fold_mode"], function(require, exports, module) {
"use strict";

var oop = require("../../lib/oop");
var Range = require("../../range").Range;
var BaseFoldMode = require("./fold_mode").FoldMode;

var FoldMode = exports.FoldMode = function(commentRegex) {
    if (commentRegex) {
        this.foldingStartMarker = new RegExp(
            this.foldingStartMarker.source.replace(/\|[^|]*?$/, "|" + commentRegex.start)
        );
        this.foldingStopMarker = new RegExp(
            this.foldingStopMarker.source.replace(/\|[^|]*?$/, "|" + commentRegex.end)
        );
    }
};
oop.inherits(FoldMode, BaseFoldMode);

(function() {
    
    this.foldingStartMarker = /(\{|\[)[^\}\]]*$|^\s*(\/\*)/;
    this.foldingStopMarker = /^[^\[\{]*(\}|\])|^[\s\*]*(\*\/)/;
    this.singleLineBlockCommentRe= /^\s*(\/\*).*\*\/\s*$/;
    this.tripleStarBlockCommentRe = /^\s*(\/\*\*\*).*\*\/\s*$/;
    this.startRegionRe = /^\s*(\/\*|\/\/)#?region\b/;
    this._getFoldWidgetBase = this.getFoldWidget;
    this.getFoldWidget = function(session, foldStyle, row) {
        var line = session.getLine(row);
    
        if (this.singleLineBlockCommentRe.test(line)) {
            if (!this.startRegionRe.test(line) && !this.tripleStarBlockCommentRe.test(line))
                return "";
        }
    
        var fw = this._getFoldWidgetBase(session, foldStyle, row);
    
        if (!fw && this.startRegionRe.test(line))
            return "start"; // lineCommentRegionStart
    
        return fw;
    };

    this.getFoldWidgetRange = function(session, foldStyle, row, forceMultiline) {
        var line = session.getLine(row);
        
        if (this.startRegionRe.test(line))
            return this.getCommentRegionBlock(session, line, row);
        
        var match = line.match(this.foldingStartMarker);
        if (match) {
            var i = match.index;

            if (match[1])
                return this.openingBracketBlock(session, match[1], row, i);
                
            var range = session.getCommentFoldRange(row, i + match[0].length, 1);
            
            if (range && !range.isMultiLine()) {
                if (forceMultiline) {
                    range = this.getSectionRange(session, row);
                } else if (foldStyle != "all")
                    range = null;
            }
            
            return range;
        }

        if (foldStyle === "markbegin")
            return;

        var match = line.match(this.foldingStopMarker);
        if (match) {
            var i = match.index + match[0].length;

            if (match[1])
                return this.closingBracketBlock(session, match[1], row, i);

            return session.getCommentFoldRange(row, i, -1);
        }
    };
    
    this.getSectionRange = function(session, row) {
        var line = session.getLine(row);
        var startIndent = line.search(/\S/);
        var startRow = row;
        var startColumn = line.length;
        row = row + 1;
        var endRow = row;
        var maxRow = session.getLength();
        while (++row < maxRow) {
            line = session.getLine(row);
            var indent = line.search(/\S/);
            if (indent === -1)
                continue;
            if  (startIndent > indent)
                break;
            var subRange = this.getFoldWidgetRange(session, "all", row);
            
            if (subRange) {
                if (subRange.start.row <= startRow) {
                    break;
                } else if (subRange.isMultiLine()) {
                    row = subRange.end.row;
                } else if (startIndent == indent) {
                    break;
                }
            }
            endRow = row;
        }
        
        return new Range(startRow, startColumn, endRow, session.getLine(endRow).length);
    };
    this.getCommentRegionBlock = function(session, line, row) {
        var startColumn = line.search(/\s*$/);
        var maxRow = session.getLength();
        var startRow = row;
        
        var re = /^\s*(?:\/\*|\/\/|--)#?(end)?region\b/;
        var depth = 1;
        while (++row < maxRow) {
            line = session.getLine(row);
            var m = re.exec(line);
            if (!m) continue;
            if (m[1]) depth--;
            else depth++;

            if (!depth) break;
        }

        var endRow = row;
        if (endRow > startRow) {
            return new Range(startRow, startColumn, endRow, line.length);
        }
    };

}).call(FoldMode.prototype);

});

define("ace/mode/pegjs",["require","exports","module","ace/lib/oop","ace/mode/text","ace/mode/pegjs_highlight_rules","ace/mode/folding/cstyle"], function(require, exports, module) {
"use strict";

var oop = require("../lib/oop");
var TextMode = require("./text").Mode;
var PEGjsHighlightRules = require("./pegjs_highlight_rules").PEGjsHighlightRules;
var FoldMode = require("./folding/cstyle").FoldMode;

var Mode = function() {
    this.HighlightRules = PEGjsHighlightRules;
    this.foldingRules = new FoldMode();
};
oop.inherits(Mode, TextMode);

(function() {
    this.$id = "ace/mode/pegjs"
}).call(Mode.prototype);

exports.Mode = Mode;
});
